### ISSUE TITLE
Overload not properly producing omitted types

### ISSUE THREAD

**Issue by njriasan** [numba/numba#8047]
## Reporting a bug

- [ ] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/main/CHANGE_LOG).
- [ ] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

It seems like the compilation path for overloads doesn't properly do argument folding and therefore doesn't produce omitted types. Here is a reproducer:

```python
from numba.extending import overload
from numba import njit

def foo(a, b=1):
	pass

@overload(foo)
def foo_overload(a, b=1):
	print(a, b)
	def foo_impl(a, b=1):
		return a + b

@njit
def g(a):
	return foo(a)

g(100)
```
On master this outputs:

```python
int64 1
Traceback (most recent call last):
  File "/Users/nicholasriasanovsky/Documents/bodo/numba/test_ex.py", line 17, in <module>
    g(100)
  File "/Users/nicholasriasanovsky/Documents/bodo/numba/numba/core/dispatcher.py", line 468, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/Users/nicholasriasanovsky/Documents/bodo/numba/numba/core/dispatcher.py", line 409, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<function foo at 0x1013af0a0>) found for signature:
 
 >>> foo(int64)
 
There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload of function 'foo': File: test_ex.py: Line 7.
    With argument(s): '(int64)':
   No match.

During: resolving callee type: Function(<function foo at 0x1013af0a0>)
During: typing of call at /Users/nicholasriasanovsky/Documents/bodo/numba/test_ex.py (15)


File "test_ex.py", line 15:
def g(a):
	return foo(a)
 ^
```

So 1 is not being converted to an omitted type.

**Comment by njriasan** [comment]
This seems to be possible to convert to a literal type if you reuse the argument folding used by dispatchers. Here is a diff that outputs `int64 omitted(default=1)`

```
diff --git a/numba/core/typing/templates.py b/numba/core/typing/templates.py
index 222be0c61..6ddd61a0d 100644
--- a/numba/core/typing/templates.py
+++ b/numba/core/typing/templates.py
@@ -790,7 +790,21 @@ class _OverloadFunctionTemplate(AbstractTemplate):
             # problems
             raise TypingError(str(e)) from e
         else:
-            ovf_result = self._overload_func(*args, **kws)
+            def normal_handler(index, param, value):
+                return value
+
+            def default_handler(index, param, default):
+                return types.Omitted(default)
+
+            def stararg_handler(index, param, values):
+                return types.StarArgTuple(values)
+
+            pysig = utils.pysignature(self._overload_func)
+            args = fold_arguments(pysig, args, kws,
+                                normal_handler,
+                                default_handler,
+                                stararg_handler)
+            ovf_result = self._overload_func(*args)
 
         if ovf_result is None:
             # No implementation => fail typing
```
Of course the python signature should probably be computed only once. Is there any reason this isn't an omitted type right now? It seems like if there was ever a default value that was unhashable then it wouldn't be possible to cache the overload.

**Comment by njriasan** [comment]
If my understanding of this is correct, I'm happy to open a PR to change this.

**Comment by esc** [comment]
@njriasan thank you for looking into this. I can confirm I am able to reproduce given the original reproducer. I can also confirm that the patch you suggested does indeed have the intended consequence. Unless there are any objections from @stuartarchibald and/or @sklam I would suggest submitting a PR with the patch and fleshing out some test cases. 👍

**Comment by stuartarchibald** [comment]
Suggest getting this in first. https://github.com/numba/numba/pull/6988

Am also not sure of the knock-on effect of changing this as IIRC it's been like this since it was first written, I wonder if there's code internally or otherwise that relies on it.

**Comment by ehsantn** [comment]
I think just using literal types instead of omitted can simplify overloads and make them much more robust. This approach loses a bit of info on whether the user has provided the value or it's the default, but I've never found this bit useful. Any thoughts?

**Comment by njriasan** [comment]
@ehsantn I believe there is at least one open issue that wants better caching with omitted types in general. I believe the desired behavior would be when compiling with literals, the Omitted type should be a literal type (assuming one exists, otherwise the Omitted type is probably still needed) and when compiling without literals the argument should be converted to the non-literal type. Not sure if this behavior should be targeted in a separate PR.

@stuartarchibald I imagine there must be code internally or elsewhere that at a minimum conforms to this behavior if not depends on it. I suspect this also exists for other APIs (overload_method at a minimum) and so there may be a lot of impacted code. I'm happy to open a PR and start working on this, but I suspect the earliest milestone that would be reasonable is 0.57.

**Comment by stuartarchibald** [comment]
> @stuartarchibald I imagine there must be code internally or elsewhere that at a minimum conforms to this behavior if not depends on it. I suspect this also exists for other APIs (overload_method at a minimum) and so there may be a lot of impacted code. I'm happy to open a PR and start working on this, but I suspect the earliest milestone that would be reasonable is 0.57.

Temporarily focusing solely on the impact of making any of the changes suggested... given this is public API and it has always behaved like this, I imagine there's a high potential for a lot of code/documentation both internally and externally to be relying on this behaviour. As a result of this I'm of the view that any change will need to go through a relatively long deprecation cycle so as to help accommodate projects using Numba that have less frequent and/or out of phase with Numba release cycles.

I suggest raising this issue at a Numba public meeting. Some things that might be good to think about:
* Whether the behaviour should change.
* How to assess the internal/external impact of changing the behaviour.
* How the behaviour could be changed and how to ensure this description is complete and relatively future proof.
* How to enforce and validate the new behaviour.
* Whether the extension API behaviours be made consistent (the public facing API needs/benefits from consistency).
* How to communicate the change to users, including how to enforce compliance of existing and new code/documentation.
* When such a change could be made with view of minimising the impact given there's a large refactoring backlog for Numba's own overload internal implementations already.
* If "bridging" code can be written to permit both old and new behaviours thus permitting a transition period.
